### ISSUE TITLE
coverity: libflux/conf defects

### ISSUE THREAD

**Issue by grondo** [flux-framework/flux-core#359]
Another set that may be false positives, but I didn't take the time to probe yet.

```
* CID 129811:  Resource leaks  (RESOURCE_LEAK)
/src/common/libflux/conf.c: 321 in flux_conf_environment_unset()


________________________________________________________________________________________________________
*** CID 129811:  Resource leaks  (RESOURCE_LEAK)
/src/common/libflux/conf.c: 321 in flux_conf_environment_unset()
315     {
316         flux_conf_environment_set_inner (cf, key, sdsnew (value), separator);
317     }
318
319     void flux_conf_environment_unset (flux_conf_t cf, const char *key)
320     {
>>>     CID 129811:  Resource leaks  (RESOURCE_LEAK)
>>>     Failing to save or free storage allocated by "sdsempty()" leaks it.
321         flux_conf_environment_set_inner (cf, key, sdsempty (), "");
322     }
323
324     static void flux_conf_environment_push_inner (flux_conf_t cf,
325                                                   const char *key,
326                                                   const char *value,

** CID 129810:    (RESOURCE_LEAK)
/src/common/libflux/conf.c: 346 in flux_conf_environment_push_inner()
/src/common/libflux/conf.c: 346 in flux_conf_environment_push_inner()
/src/common/libflux/conf.c: 346 in flux_conf_environment_push_inner()


________________________________________________________________________________________________________
*** CID 129810:    (RESOURCE_LEAK)
/src/common/libflux/conf.c: 346 in flux_conf_environment_push_inner()
340             if (before) {
341                 usa_push (item, value);
342             } else {
343                 usa_push_back (item, value);
344             }
345         }
>>>     CID 129810:    (RESOURCE_LEAK)
>>>     Variable "item" going out of scope leaks the storage it points to.
346     }
347
348     void flux_conf_environment_push (flux_conf_t cf,
349                                              const char *key,
350                                              const char *value)
351     {
/src/common/libflux/conf.c: 346 in flux_conf_environment_push_inner()
340             if (before) {
341                 usa_push (item, value);
342             } else {
343                 usa_push_back (item, value);
344             }
345         }
>>>     CID 129810:    (RESOURCE_LEAK)
>>>     Variable "item" going out of scope leaks the storage it points to.
346     }
347
348     void flux_conf_environment_push (flux_conf_t cf,
349                                              const char *key,
350                                              const char *value)
351     {
/src/common/libflux/conf.c: 346 in flux_conf_environment_push_inner()
340             if (before) {
341                 usa_push (item, value);
342             } else {
343                 usa_push_back (item, value);
344             }
345         }
>>>     CID 129810:    (RESOURCE_LEAK)
>>>     Variable "item" going out of scope leaks the storage it points to.
346     }
347
348     void flux_conf_environment_push (flux_conf_t cf,
349                                              const char *key,
350                                              const char *value)
351     {

** CID 129809:  Resource leaks  (RESOURCE_LEAK)
/src/common/libflux/conf.c: 316 in flux_conf_environment_set()


________________________________________________________________________________________________________
*** CID 129809:  Resource leaks  (RESOURCE_LEAK)
/src/common/libflux/conf.c: 316 in flux_conf_environment_set()
310
311     void flux_conf_environment_set (flux_conf_t cf,
312                                     const char *key,
313                                     const char *value,
314                                     const char *separator)
315     {
>>>     CID 129809:  Resource leaks  (RESOURCE_LEAK)
>>>     Failing to save or free storage allocated by "sdsnew(value)" leaks it.
316         flux_conf_environment_set_inner (cf, key, sdsnew (value), separator);
317     }
318
319     void flux_conf_environment_unset (flux_conf_t cf, const char *key)
320     {
321         flux_conf_environment_set_inner (cf, key, sdsempty (), "");

```

**Comment by trws** [comment]
I'm pretty sure those are all false positives.  Here's hoping for ASan soon... anyway, the set_inner function expects an allocated `sds` as an argument, and adds it to the `usa` structure where it is freed either when it is replaced or when the usa itself is destroyed.  The `item` leaks are functions that get the pointer out of the hash table as a non-owned reference.  Those are all managed by the hash table.

**Comment by dongahn** [comment]
The ASan page says there should be no problem in supporting shared libraries so I will take a look at this except: 

>
Q: When I link my shared library with -fsanitize=address, it fails due to some undefined ASan symbols (e.g. asan_init_v4)?
A: Most probably you link with -Wl,-z,defs or -Wl,--no-undefined. These flags don't work with ASan.

If worse comes to the worst, I can always go to Kostya :-) 

Also, when ASan finds a memory error, it stops the execution and return -1 so this can be integrated with the CI although i will have to check whether the presence of a leak will cause the executable to return a negative.

**Comment by trws** [comment]
Yeah, I made a branch that had a `--with-sanitizer=<san>` flag on configure that had the side-effect of turning off `-Wl,--no-undefined`, but ran into the other issues before getting it to a point where it was worth putting up.  It should be pretty simple otherwise, just have to ensure that clang is used, but then we can hook it just about anywhere by just picking a config option.

**Comment by dongahn** [comment]
Ok. How can I get your patch (the config option + turning off `-Wl,--no-defined`) so that I don't have to duplicate the effort? I would like to narrow down this problem before going to Kostya.

**Comment by trws** [comment]
Just now pushed it up, you can grab it from my [sanitizers branch](https://github.com/trws/flux-core/tree/sanitizers).

**Comment by dongahn** [comment]
Got it. Thanks.

**Comment by dongahn** [comment]
OK, I've made some progress. Several things:

1. @trws: was the reason that you couldn't get clang-version to work with -static-libstdc++ or something else? https://code.google.com/p/address-sanitizer/issues/detail?id=147

In any case, you are right: clang doesn't not support shared library ASAN although we can do that if we build a customized clang as documented in https://code.google.com/p/address-sanitizer/wiki/AsanAsDso

2. Oddly enough, gcc (higher or equal to 4.8) does support shared library ASAN (! competition?) and I was able to build flux-core with it after working around several bumps. One of the bumps was -Wl,--no-undefined. I have manually disabled those to see how far I can do and I was able to go until  completion. But a question is: is there a central place you can disable this linker flag  when sanitizer config flag is on or do you have to change many Makefile.am files? 

3. Even with gcc, I think we will have to use this in a targeted way because SAN doesn't support some low-level mechanism we are using. In fact, `make check` stopped with the following messages. 

>
==171909== WARNING: ASan doesn't fully support makecontext/swapcontext functions and may produce false positives in some cases!

>
==171909== ERROR: AddressSanitizer: SEGV on unknown address 0x2aaaafb8a008 (pc 0x2aaaaacd509d sp 0x7ffffffe9b30 bp 0x2aaaafb8a000 T0)

You can ignore certain functions for ASAN instrumentation with no_sanitize_address attribute (__attribute__((no_sanitize_address))). You can also do this with black list capability. So it seems we have some knobs we can play wiht. 

Is there any tests that I can manually run to get a bit more feeling of this technology, which of course won't use coproc?

**Comment by trws** [comment]
It's possible that's the reason, but I didn't specify that.  It just wouldn't link the santitizer symbols into our shared libraries.

As to the makecontext/swapcontext stuff, that could be a problem... I don't know how many tests we have that can go around that, but I'm surprised that it would segfault as a result.  At the least, most of the non-session tests should run, the ones under src I mean.

**Comment by trws** [comment]
Come to think of it, @garlick, is there a reasonable way to switch coprocesses off anymore? I remember you had a flag for it in configure at one point.

**Comment by dongahn** [comment]
@trws I actually made both clang and gcc work and found lots of leaks (need some analysis) I will make this version available soon at my fork and let you folks try.

**Comment by trws** [comment]
Great! Well... not so great about the leaks, but great that we have a good way to find them. =)

**Comment by dongahn** [comment]
Yes, I am excited about this. I also briefly tried TSan and it seemed to find a race as well. Good ways to improve code quality going forward.

**Comment by garlick** [comment]
No easy way to compile without makecontext and friends, though it's abstracted in libutil/coproc.c.

**Comment by dongahn** [comment]
I think we can get good mileage without turning this off. BTW, is there any option that allow `make check` continue even if a failure is detected? ASan report any memory error as a fatal error and as such even a leak stops `make check` So I manually cd into each sub-directory and ran some tests to see if we can get some good diagnostics.

**Comment by grondo** [comment]
@dongahn, you could try `make -i check`

**Comment by dongahn** [comment]
Got it. Thanks.

**Comment by trws** [comment]
I actually meant run without invoking them, rather than build without them.  ASan shouldn't even notice they're there unless they're called.  Did the FLUX_O_COPROC switch go away?

**Comment by trws** [comment]
If it's sufficiently worthwhile by the way, I think our coprocesses represent a non-cyclic set anyway, such that they should execute correctly, if slowly, without actually yielding to the reactor on blocking.  Even if not, it would be relatively straightforward to implement a secondary version that implements the same interface on top of pthreads.

**Comment by garlick** [comment]
Sorry, FLUX_O_COPROC is there and I believe it is only set in a few tests, we've not really used this in anger yet.  It is however somewhat vital to the correctness of some loop tests that answer their own requests.

**Comment by trws** [comment]
Thanks @garlick, I'm sorry for not being more precise initially.  That being the case, I could bang up an alternative pthreads version that conforms to the same interface if you like, either that would run in parallel or that would ensure the exact same ordering as now by blocking all threads that are not currently in the "running" state as the coprocs do.

**Comment by garlick** [comment]
Would it be better to find some way to skip the tests that trip up the analysis tool for now?

It seems like a lot of work to create (and validate, maintain) and alternate version of coprocs.

**Comment by dongahn** [comment]
Like I said, it seems we can have good mileage as it for now. I will soon make my branch available for you folks. As a teaser, here are some of the reports I am getting. It seems CLANG and GCC found some distinct sets of issues (as the memory layout will be different between two compilers). I think the diagnostics should look better when `-O0 -g` is given, but I didn't try this. 

## Leak report example from ASan (Clang)

```
=================================================================
==214147==ERROR: LeakSanitizer: detected memory leaks

    #0 0x49da39 in malloc /collab/usr/global/tools/clang/chaos_5_x86_64_ib/clang-omp-3.5.0/src/llvm-3.5.0.src/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:40
Direct leak of 663 byte(s) in 10 object(s) allocated from:
    #1 0x2aaaab200b87 in sdsnewlen /g/g0/dahn/workspace/ASAN/dongahn_fork/flux-core/build/src/common/libutil/../../../../src/common/libutil/sds.c:91:10

Direct leak of 528 byte(s) in 6 object(s) allocated from:
    #0 0x49da39 in malloc /collab/usr/global/tools/clang/chaos_5_x86_64_ib/clang-omp-3.5.0/src/llvm-3.5.0.src/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:40
    #0 0x49da39 in malloc /collab/usr/global/tools/clang/chaos_5_x86_64_ib/clang-omp-3.5.0/src/llvm-3.5.0.src/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:40
    #1 0x2aaaab200b87 in sdsnewlen /g/g0/dahn/workspace/ASAN/dongahn_fork/flux-core/build/src/common/libutil/../../../../src/common/libutil/sds.c:91:10

Direct leak of 528 byte(s) in 6 object(s) allocated from:
    #1 0x2aaaabb39cb0 (/usr/lib64/libzmq.so.4+0x1dcb0)

Direct leak of 400 byte(s) in 10 object(s) allocated from:
    #0 0x49da39 in malloc /collab/usr/global/tools/clang/chaos_5_x86_64_ib/clang-omp-3.5.0/src/llvm-3.5.0.src/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:40
    #0 0x49da39 in malloc /collab/usr/global/tools/clang/chaos_5_x86_64_ib/clang-omp-3.5.0/src/llvm-3.5.0.src/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:40
    #1 0x2aaaabb39cb0 (/usr/lib64/libzmq.so.4+0x1dcb0)

Direct leak of 400 byte(s) in 10 object(s) allocated from:
    #1 0x2aaaacaaf45f in _IO_vasprintf /usr/src/debug/glibc-2.12-2-gc4ccff1/libio/vasprintf.c:78

Direct leak of 144 byte(s) in 2 object(s) allocated from:
    #0 0x49da39 in malloc /collab/usr/global/tools/clang/chaos_5_x86_64_ib/clang-omp-3.5.0/src/llvm-3.5.0.src/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:40
    #0 0x49dba9 in calloc /collab/usr/global/tools/clang/chaos_5_x86_64_ib/clang-omp-3.5.0/src/llvm-3.5.0.src/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:56
    #1 0x2aaaacaaf45f in _IO_vasprintf /usr/src/debug/glibc-2.12-2-gc4ccff1/libio/vasprintf.c:78

Direct leak of 144 byte(s) in 2 object(s) allocated from:
    #1 0x2aaaab694871 in json_object_new /builddir/build/BUILD/json-c-json-c-0.11-20130402/json_object.c:178
    #2 0x2aaab4013197 in wait_runone /g/g0/dahn/workspace/ASAN/dongahn_fork/flux-core/build/src/modules/kvs/../../../../src/modules/kvs/waitqueue.c:137:9
    #3 0x2aaab4013197 in wait_runqueue /g/g0/dahn/workspace/ASAN/dongahn_fork/flux-core/build/src/modules/kvs/../../../../src/modules/kvs/waitqueue.c:157
    #0 0x49dba9 in calloc /collab/usr/global/tools/clang/chaos_5_x86_64_ib/clang-omp-3.5.0/src/llvm-3.5.0.src/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:56
    #4 0x2aaab400bea4 in load_complete /g/g0/dahn/workspace/ASAN/dongahn_fork/flux-core/build/src/modules/kvs/../../../../src/modules/kvs/kvs.c:382:5
    #5 0x2aaab400bea4 in load_response_cb /g/g0/dahn/workspace/ASAN/dongahn_fork/flux-core/build/src/modules/kvs/../../../../src/modules/kvs/kvs.c:778
    #1 0x2aaaab694871 in json_object_new /builddir/build/BUILD/json-c-json-c-0.11-20130402/json_object.c:178
    #6 0x4e0ac8 in msg_compat_cb /g/g0/dahn/workspace/ASAN/dongahn_fork/flux-core/build/src/common/libcompat/../../../../src/common/libcompat/reactor.c:137:9
    #2 0x2aaab4013197 in wait_runone /g/g0/dahn/workspace/ASAN/dongahn_fork/flux-core/build/src/modules/kvs/../../../../src/modules/kvs/waitqueue.c:137:9
    #3 0x2aaab4013197 in wait_runqueue /g/g0/dahn/workspace/ASAN/dongahn_fork/flux-core/build/src/modules/kvs/../../../../src/modules/kvs/waitqueue.c:157
    #7 0x2aaaab1ad5df in handle_cb /g/g0/dahn/workspace/ASAN/dongahn_fork/flux-core/build/src/common/libflux/../../../../src/common/libflux/reactor.c:361:13
    #4 0x2aaab400bea4 in load_complete /g/g0/dahn/workspace/ASAN/dongahn_fork/flux-core/build/src/modules/kvs/../../../../src/modules/kvs/kvs.c:382:5
    #5 0x2aaab400bea4 in load_response_cb /g/g0/dahn/workspace/ASAN/dongahn_fork/flux-core/build/src/modules/kvs/../../../../src/modules/kvs/kvs.c:778
    #6 0x4e0ac8 in msg_compat_cb /g/g0/dahn/workspace/ASAN/dongahn_fork/flux-core/build/src/common/libcompat/../../../../src/common/libcompat/reactor.c:137:9
    #8 0x2aaaab21284e in ev_invoke_pending /g/g0/dahn/workspace/ASAN/dongahn_fork/flux-core/build/src/common/libev/../../../../src/common/libev/ev.c:3088:11
    #9 0x2aaaab1ab442 in flux_reactor_start /g/g0/dahn/workspace/ASAN/dongahn_fork/flux-core/build/src/common/libflux/../../../../src/common/libflux/reactor.c:127:5
    #10 0x4c7ac4 in module_thread /g/g0/dahn/workspace/ASAN/dongahn_fork/flux-core/build/src/broker/../../../src/broker/module.c:133:9
    #7 0x2aaaab1ad5df in handle_cb /g/g0/dahn/workspace/ASAN/dongahn_fork/flux-core/build/src/common/libflux/../../../../src/common/libflux/reactor.c:361:13
    #11 0x2aaaabd7f9d0 in start_thread /usr/src/debug/glibc-2.12-2-gc4ccff1/nptl/pthread_create.c:301

Direct leak of 81 byte(s) in 10 object(s) allocated from:
    #8 0x2aaaab21284e in ev_invoke_pending /g/g0/dahn/workspace/ASAN/dongahn_fork/flux-core/build/src/common/libev/../../../../src/common/libev/ev.c:3088:11
    #9 0x2aaaab1ab442 in flux_reactor_start /g/g0/dahn/workspace/ASAN/dongahn_fork/flux-core/build/src/common/libflux/../../../../src/common/libflux/reactor.c:127:5
    #10 0x4c7ac4 in module_thread /g/g0/dahn/workspace/ASAN/dongahn_fork/flux-core/build/src/broker/../../../src/broker/module.c:133:9
    #0 0x49da39 in malloc /collab/usr/global/tools/clang/chaos_5_x86_64_ib/clang-omp-3.5.0/src/llvm-3.5.0.src/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:40
    #1 0x2aaaacac0e91 in __GI___strdup /usr/src/debug/glibc-2.12-2-gc4ccff1/string/strdup.c:43

Direct leak of 72 byte(s) in 1 object(s) allocated from:
    #11 0x2aaaabd7f9d0 in start_thread /usr/src/debug/glibc-2.12-2-gc4ccff1/nptl/pthread_create.c:301

Direct leak of 81 byte(s) in 10 object(s) allocated from:
    #0 0x49dba9 in calloc /collab/usr/global/tools/clang/chaos_5_x86_64_ib/clang-omp-3.5.0/src/llvm-3.5.0.src/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:56
    #1 0x2aaaab694871 in json_object_new /builddir/build/BUILD/json-c-json-c-0.11-20130402/json_object.c:178
    #0 0x49da39 in malloc /collab/usr/global/tools/clang/chaos_5_x86_64_ib/clang-omp-3.5.0/src/llvm-3.5.0.src/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:40
    #2 0x4e0ac8 in msg_compat_cb /g/g0/dahn/workspace/ASAN/dongahn_fork/flux-core/build/src/common/libcompat/../../../../src/common/libcompat/reactor.c:137:9
    #1 0x2aaaacac0e91 in __GI___strdup /usr/src/debug/glibc-2.12-2-gc4ccff1/string/strdup.c:43

Direct leak of 72 byte(s) in 1 object(s) allocated from:
    #3 0x2aaaab1ad5df in handle_cb /g/g0/dahn/workspace/ASAN/dongahn_fork/flux-core/build/src/common/libflux/../../../../src/common/libflux/reactor.c:361:13
    #0 0x49dba9 in calloc /collab/usr/global/tools/clang/chaos_5_x86_64_ib/clang-omp-3.5.0/src/llvm-3.5.0.src/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:56
    #4 0x2aaaab21284e in ev_invoke_pending /g/g0/dahn/workspace/ASAN/dongahn_fork/flux-core/build/src/common/libev/../../../../src/common/libev/ev.c:3088:11
    #1 0x2aaaab694871 in json_object_new /builddir/build/BUILD/json-c-json-c-0.11-20130402/json_object.c:178
    #5 0x2aaaab1ab442 in flux_reactor_start /g/g0/dahn/workspace/ASAN/dongahn_fork/flux-core/build/src/common/libflux/../../../../src/common/libflux/reactor.c:127:5
    #2 0x4e0ac8 in msg_compat_cb /g/g0/dahn/workspace/ASAN/dongahn_fork/flux-core/build/src/common/libcompat/../../../../src/common/libcompat/reactor.c:137:9
    #6 0x4c7ac4 in module_thread /g/g0/dahn/workspace/ASAN/dongahn_fork/flux-core/build/src/broker/../../../src/broker/module.c:133:9
    #3 0x2aaaab1ad5df in handle_cb /g/g0/dahn/workspace/ASAN/dongahn_fork/flux-core/build/src/common/libflux/../../../../src/common/libflux/reactor.c:361:13
    #7 0x2aaaabd7f9d0 in start_thread /usr/src/debug/glibc-2.12-2-gc4ccff1/nptl/pthread_create.c:301

Direct leak of 32 byte(s) in 2 object(s) allocated from:
    #4 0x2aaaab21284e in ev_invoke_pending /g/g0/dahn/workspace/ASAN/dongahn_fork/flux-core/build/src/common/libev/../../../../src/common/libev/ev.c:3088:11
    #5 0x2aaaab1ab442 in flux_reactor_start /g/g0/dahn/workspace/ASAN/dongahn_fork/flux-core/build/src/common/libflux/../../../../src/common/libflux/reactor.c:127:5
    #0 0x49dba9 in calloc /collab/usr/global/tools/clang/chaos_5_x86_64_ib/clang-omp-3.5.0/src/llvm-3.5.0.src/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:56
    #1 0x4e0801 in cleanup_push /g/g0/dahn/workspace/ASAN/dongahn_fork/flux-core/build/src/common/libutil/../../../../src/common/libutil/cleanup.c:89:26
    #6 0x4c7ac4 in module_thread /g/g0/dahn/workspace/ASAN/dongahn_fork/flux-core/build/src/broker/../../../src/broker/module.c:133:9

Direct leak of 16 byte(s) in 1 object(s) allocated from:
    #7 0x2aaaabd7f9d0 in start_thread /usr/src/debug/glibc-2.12-2-gc4ccff1/nptl/pthread_create.c:301

Direct leak of 32 byte(s) in 2 object(s) allocated from:
    #0 0x49dba9 in calloc /collab/usr/global/tools/clang/chaos_5_x86_64_ib/clang-omp-3.5.0/src/llvm-3.5.0.src/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:56
    #1 0x4e0801 in cleanup_push /g/g0/dahn/workspace/ASAN/dongahn_fork/flux-core/build/src/common/libutil/../../../../src/common/libutil/cleanup.c:89:26
    #2 0x2aaaaca5ed5c in __libc_start_main /usr/src/debug/glibc-2.12-2-gc4ccff1/csu/libc-start.c:226

Direct leak of 16 byte(s) in 1 object(s) allocated from:
    #0 0x49dba9 in calloc /collab/usr/global/tools/clang/chaos_5_x86_64_ib/clang-omp-3.5.0/src/llvm-3.5.0.src/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:56
    #1 0x4e0801 in cleanup_push /g/g0/dahn/workspace/ASAN/dongahn_fork/flux-core/build/src/common/libutil/../../../../src/common/libutil/cleanup.c:89:26

Direct leak of 16 byte(s) in 1 object(s) allocated from:
    #0 0x49dba9 in calloc /collab/usr/global/tools/clang/chaos_5_x86_64_ib/clang-omp-3.5.0/src/llvm-3.5.0.src/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:56
    #1 0x2aaab28d3861 in cleanup_push /g/g0/dahn/workspace/ASAN/dongahn_fork/flux-core/build/src/common/libutil/../../../../src/common/libutil/cleanup.c:89:26
    #2 0x4c7ac4 in module_thread /g/g0/dahn/workspace/ASAN/dongahn_fork/flux-core/build/src/broker/../../../src/broker/module.c:133:9
    #0 0x49dba9 in calloc /collab/usr/global/tools/clang/chaos_5_x86_64_ib/clang-omp-3.5.0/src/llvm-3.5.0.src/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:56
    #1 0x4e0801 in cleanup_push /g/g0/dahn/workspace/ASAN/dongahn_fork/flux-core/build/src/common/libutil/../../../../src/common/libutil/cleanup.c:89:26
    #3 0x2aaaabd7f9d0 in start_thread /usr/src/debug/glibc-2.12-2-gc4ccff1/nptl/pthread_create.c:301

Indirect leak of 1536 byte(s) in 3 object(s) allocated from:
    #2 0x2aaaaca5ed5c in __libc_start_main /usr/src/debug/glibc-2.12-2-gc4ccff1/csu/libc-start.c:226

Direct leak of 16 byte(s) in 1 object(s) allocated from:
    #0 0x49dba9 in calloc /collab/usr/global/tools/clang/chaos_5_x86_64_ib/clang-omp-3.5.0/src/llvm-3.5.0.src/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:56
    #1 0x2aaaab6982c6 in lh_table_new /builddir/build/BUILD/json-c-json-c-0.11-20130402/linkhash.c:439

Indirect leak of 264 byte(s) in 3 object(s) allocated from:
    #0 0x49dba9 in calloc /collab/usr/global/tools/clang/chaos_5_x86_64_ib/clang-omp-3.5.0/src/llvm-3.5.0.src/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:56
    #1 0x2aaab28d3861 in cleanup_push /g/g0/dahn/workspace/ASAN/dongahn_fork/flux-core/build/src/common/libutil/../../../../src/common/libutil/cleanup.c:89:26
    
== CUT ==
```



## Race report example from TSan (Clang)

```
==================
WARNING: ThreadSanitizer: data race (pid=162110)
  Read of size 8 at 0x7da000008090 by thread T1:
    #0 epoll_ctl /collab/usr/global/tools/clang/chaos_5_x86_64_ib/clang-omp-3.5.0/src/llvm-3.5.0.src/projects/compiler-rt/lib/tsan/rtl/tsan_interceptors.cc:1664 (test_message.t+0x0000000439fc)
    #1 <null> <null>:0 (libzmq.so.4+0x000000017ebf)

  Previous write of size 8 at 0x7da000008090 by main thread (mutexes: write M5):
    #0 eventfd /collab/usr/global/tools/clang/chaos_5_x86_64_ib/clang-omp-3.5.0/src/llvm-3.5.0.src/projects/compiler-rt/lib/tsan/rtl/tsan_interceptors.cc:1426 (test_message.t+0x00000002d773)
    #1 <null> <null>:0 (libzmq.so.4+0x000000030421)
    #2 main /g/g0/dahn/workspace/ASAN/dongahn_fork/flux-core/build/src/common/libflux/../../../../src/common/libflux/test/message.c:485:5 (test_message.t+0x00000009c13a)

  Location is file descriptor 9 created by main thread at:
    #0 eventfd /collab/usr/global/tools/clang/chaos_5_x86_64_ib/clang-omp-3.5.0/src/llvm-3.5.0.src/projects/compiler-rt/lib/tsan/rtl/tsan_interceptors.cc:1426 (test_message.t+0x00000002d773)
    #1 <null> <null>:0 (libzmq.so.4+0x000000030421)
    #2 main /g/g0/dahn/workspace/ASAN/dongahn_fork/flux-core/build/src/common/libflux/../../../../src/common/libflux/test/message.c:485:5 (test_message.t+0x00000009c13a)

  Mutex M5 (0x7d500000fe40) created at:
    #0 pthread_mutex_init /collab/usr/global/tools/clang/chaos_5_x86_64_ib/clang-omp-3.5.0/src/llvm-3.5.0.src/projects/compiler-rt/lib/tsan/rtl/tsan_interceptors.cc:1032 (test_message.t+0x000000038635)
    #1 <null> <null>:0 (libzmq.so.4+0x00000000e9ab)
    #2 main /g/g0/dahn/workspace/ASAN/dongahn_fork/flux-core/build/src/common/libflux/../../../../src/common/libflux/test/message.c:485:5 (test_message.t+0x00000009c13a)

  Thread T1 (tid=162112, running) created by main thread at:
    #0 pthread_create /collab/usr/global/tools/clang/chaos_5_x86_64_ib/clang-omp-3.5.0/src/llvm-3.5.0.src/projects/compiler-rt/lib/tsan/rtl/tsan_interceptors.cc:848 (test_message.t+0x00000004c183)
    #1 <null> <null>:0 (libzmq.so.4+0x00000003d85c)
    #2 main /g/g0/dahn/workspace/ASAN/dongahn_fork/flux-core/build/src/common/libflux/../../../../src/common/libflux/test/message.c:485:5 (test_message.t+0x00000009c13a)

SUMMARY: ThreadSanitizer: data race ??:0 ??
==================
ThreadSanitizer: reported 1 warnings

```


## Memory error report example from ASan (GCC)

```
==46300== ERROR: AddressSanitizer: heap-buffer-overflow on address 0x60080000be38 at pc 0x2aaaae14f9ad bp 0x7fffffffcc70 sp 0x7fffffffcc68
READ of size 8 at 0x60080000be38 thread T0
    #0 0x2aaaae14f9ac (/g/g0/dahn/workspace/ASAN/flux-core/gcc-4.8.2.p-san/src/common/.libs/libflux-core.so.0.0.0+0x609ac)
    #1 0x2aaaae1196ab (/g/g0/dahn/workspace/ASAN/flux-core/gcc-4.8.2.p-san/src/common/.libs/libflux-core.so.0.0.0+0x2a6ab)
    #2 0x2aaaae11a5e5 (/g/g0/dahn/workspace/ASAN/flux-core/gcc-4.8.2.p-san/src/common/.libs/libflux-core.so.0.0.0+0x2b5e5)
    #3 0x2aaaae11a7cf (/g/g0/dahn/workspace/ASAN/flux-core/gcc-4.8.2.p-san/src/common/.libs/libflux-core.so.0.0.0+0x2b7cf)
    #4 0x40d863 (/g/g0/dahn/workspace/ASAN/flux-core/gcc-4.8.2.p-san/src/broker/.libs/lt-flux-broker+0x40d863)
    #5 0x2aaaaf2fdd5c (/lib64/libc-2.12.so+0x1ed5c)
    #6 0x407e9c (/g/g0/dahn/workspace/ASAN/flux-core/gcc-4.8.2.p-san/src/broker/.libs/lt-flux-broker+0x407e9c)
0x60080000be38 is located 0 bytes to the right of 40-byte region [0x60080000be10,0x60080000be38)
allocated by thread T0 here:
    #0 0x2aaaaace255a (/collab/usr/gapps/opnsrc/gnu/dev/lnx-2.12-o/4.8.2/lib64/libasan.so.0.0.0+0x1555a)
    #1 0x2aaaae14cad4 (/g/g0/dahn/workspace/ASAN/flux-core/gcc-4.8.2.p-san/src/common/.libs/libflux-core.so.0.0.0+0x5dad4)
    #2 0x2aaaae14f711 (/g/g0/dahn/workspace/ASAN/flux-core/gcc-4.8.2.p-san/src/common/.libs/libflux-core.so.0.0.0+0x60711)
    #3 0x2aaaae1196ab (/g/g0/dahn/workspace/ASAN/flux-core/gcc-4.8.2.p-san/src/common/.libs/libflux-core.so.0.0.0+0x2a6ab)
    #4 0x2aaaae11a5e5 (/g/g0/dahn/workspace/ASAN/flux-core/gcc-4.8.2.p-san/src/common/.libs/libflux-core.so.0.0.0+0x2b5e5)
    #5 0x2aaaae11a7cf (/g/g0/dahn/workspace/ASAN/flux-core/gcc-4.8.2.p-san/src/common/.libs/libflux-core.so.0.0.0+0x2b7cf)
    #6 0x40d863 (/g/g0/dahn/workspace/ASAN/flux-core/gcc-4.8.2.p-san/src/broker/.libs/lt-flux-broker+0x40d863)
    #7 0x2aaaaf2fdd5c (/lib64/libc-2.12.so+0x1ed5c)
Shadow bytes around the buggy address:
  0x0c017fff9770: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c017fff9780: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c017fff9790: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c017fff97a0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c017fff97b0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
=>0x0c017fff97c0: fa fa 00 00 00 00 00[fa]fa fa 00 00 00 00 00 fa
  0x0c017fff97d0: fa fa 00 00 00 00 00 00 fa fa 00 00 00 00 00 00
  0x0c017fff97e0: fa fa 00 00 00 00 00 00 fa fa 00 00 00 00 00 fa
  0x0c017fff97f0: fa fa 00 00 00 00 00 00 fa fa 00 00 00 00 00 00
  0x0c017fff9800: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c017fff9810: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:     fa
  Heap righ redzone:     fb
  Freed Heap region:     fd
  Stack left redzone:    f1
  Stack mid redzone:     f2
  Stack right redzone:   f3
  Stack partial redzone: f4
  Stack after return:    f5
  Stack use after scope: f8
  Global redzone:        f9
  Global init order:     f6
  Poisoned by user:      f7
  ASan internal:         fe
==46300== ABORTING
flux-start: 1 (pid 46300) exited with rc=1
=================================================================

```

**Comment by dongahn** [comment]
The initial Sanitizer support is now available at [the san branch](https://github.com/dongahn/flux-core/tree/san). I only tested this on LC cab system with `--disable-python`. If you want to build a sanitized flux using clang, at the root of the source tree, do the following:

```
use clang-omp-3.5.0
mkdir clang-san
cd clang-san
setenv ASAN_OPTIONS detect_odr_violation=0
../configure --disable-python --enable-sanitizer=address CC=clang
make
make -i check
```

I recommend to set SAN_OPTIONS that way so that [ODR violation](https://code.google.com/p/address-sanitizer/wiki/OneDefinitionRuleViolation) won't stop the building. 

If you want to build a sanitized flux using gcc, similarly at the root of the source tree, do the following:
```
use gcc-4.8.2p
mkdir gcc4.8.2.p-san
cd gcc4.8.2.p-san
../configure --disable-python --enable-sanitizer=address
setenv ASAN_OPTIONS detect_odr_violation=0
make
make -i check
```

Many tests will fail because of sanitizer checks; we probably want to look at one San issue at a time. 

If you need further suggestions and San support, let me know. Also if you folks want to merge this into the main, let me know also. I will create a PR at which point we can embellish it a bit to make it more worthwhile to be merged.

**Comment by garlick** [comment]
libflux/conf is stripped down to practically nothing.  Closing as the problems described are for code that is gone now.
